lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java

```diff
@@ -1649,6 +1649,125 @@ public void onPresenceMessage(PresenceMessage message) {
+
+        try {
+            ClientOptions option1 = createOptions(testVars.keys[0].keyStr);
+            option1.clientId = "emitter client";
+            ClientOptions option2 = createOptions(testVars.keys[0].keyStr);
+            option2.clientId = "receiver client";
+
+            ably1 = new AblyRealtime(option1);
+            ably2 = new AblyRealtime(option2);
+
+            Channel channel1 = ably1.channels.get(channelName);
+
+            Channel channel2 = ably2.channels.get(channelName);
+            channel2.attach();
+            (new ChannelWaiter(channel2)).waitFor(ChannelState.attached);
+
+            CountingCompletionListener messageCompletionListener = new CountingCompletionListener();
+
+            final ArrayList<PresenceMessage> receivedMessageStack = new ArrayList<>();
+            channel2.presence.subscribe(actions, new Presence.PresenceListener() {
+                @Override
+                public void onPresenceMessage(PresenceMessage message) {
+                    synchronized (receivedMessageStack) {
+                        receivedMessageStack.add(message);
+                        receivedMessageStack.notify();
+                    }
+                }
+            });
+
+            /* Start emitting channel with ably client 1 (emitter) */
+            channel1.presence.enter("Hello, #2!", messageCompletionListener);
+            channel1.presence.update("Lorem ipsum", messageCompletionListener);
+            channel1.presence.update("Dolor sit!", messageCompletionListener);
+
+            /* Wait until receiver client (ably2) observes {@code Action.leave}
+             * is emitted from emitter client (ably1)
+             */
+            try {
+                synchronized (receivedMessageStack) {
+                    while (receivedMessageStack.size() == 0 ||
+                            !receivedMessageStack.get(receivedMessageStack.size()-1).clientId.equals(ably1.options.clientId) ||
+                            !receivedMessageStack.get(receivedMessageStack.size()-1).data.equals("Dolor sit!"))
+                                receivedMessageStack.wait();
+                }
+            } catch(InterruptedException e) {}
+
+            /* Validate that,
+             *- we received specific actions
+             */
+            assertThat(receivedMessageStack.size(), is(equalTo(3)));
+            for (PresenceMessage message : receivedMessageStack) {
+                assertTrue(actions.contains(message.action));
+            }
+
+            /*
+             * Validate that
+             * - our listeners are called
+             */
+            try {
+                while (true) {
+                    synchronized (messageCompletionListener) {
+                        assertEquals(0, messageCompletionListener.failedListeners);
+
+                        if (messageCompletionListener.successfulListeners != 3) {
+                            messageCompletionListener.wait(5000);
```

Comment:
Isn't this going to loop forever if we don't reach `messageCompletionListener.successfulListeners == 3` ?

Reply:
Ah, yes - you're right. I mistakenly thought the timeout would throw an exception if the timeout was exceeded. I will add a check.

Reply:
It turns out there's an existing listener that kinda already does this - I've adapted it to have timeouts and used that :)